### PR TITLE
Add unique index on entityType 'name' column

### PR DESCRIPTION
**Pull request by colemanw**
- This index ensures data integrity, keeping 'name' unique.
- It also lets the system know that 'name' can be derived from 'label' if not provided.
- Adds the first (very simple) working unit test to this extension, demonstrating that name can be optional.
- Regenerates the boilerplate DAO and SQL files.

### CODE REVIEW

**Comment by colemanw** [comment]
@jensschuppe I'm aware that APIv4 is still broken, as creating an EckEntityType does not actually run the `\CRM_Eck_BAO_EckEntityType::ensureEntityType()` which creates the tables and option groups, etc. However, this is a step in the right direction and adds a simple unit test which can be built upon to get APIv4 fully working.

Once this is merged, I think the next step is to stop calling `ensureEntityType()` from the form layer, and move it to a post hook in `CRM_Eck_BAO_EckEntityType` so that it always runs.

File: CRM/Eck/DAO/EckEntityType.php [hunk 1]
```diff
@@ -238,63 +203,18 @@ public static function &export($prefix = FALSE) {
...
-    }
-
-    // Flush schema cache.
-    CRM_Core_DAO_AllCoreTables::reinitializeCache();
-
-    // Flush navigation cache.
-    CRM_Core_BAO_Navigation::resetNavigation();
-
-    return $result;
-  }
```
Comment (jensschuppe) [inline CRM/Eck/DAO/EckEntityType.php:298]: This procedure for deleting dependent stuff should then also be moved into a post hook, @colemanw? `ensureEntityType()` does not take care of an entity type being deleted.

Comment (colemanw) [reply CRM/Eck/DAO/EckEntityType.php:298]: Ah, I hadn't noticed there was hand-written code in that DAO file which got deleted by regenerating the file.
So, yes you can override the BAO deleteRecord() method and that will work. But I think a neater place to put the code would be in `hook_civicrm_post`. You can do so directly in the BAO file by implementing `HookInterface`. Here's an example: https://github.com/civicrm/civicrm-core/commit/42f0bf162b77ef27b0e93a6b2e882491cdfd8752

**Comment by colemanw** [comment]
There are merge-conflicts in this PR @jensschuppe. Give me a minute and I'll rebase this and add in the post hook.

File: CRM/Eck/DAO/EckEntityType.php (hunk 1 above)
Comment (jensschuppe) [reply CRM/Eck/DAO/EckEntityType.php:298]: Interesting, didn't know about that. Now I'm just curious about why that interface is in the `Civi\Test` namespace when it is supposed to identify classes with hook implementations for runtime.

Comment (colemanw) [reply CRM/Eck/DAO/EckEntityType.php:298]: Yea, it belongs in a better namespace for sure.

**Comment by colemanw** [comment]
@jensschuppe I think this is ready to be merged now. I improved the unit test but creating an entity type via API is still broken.
After merging this, the next thing to do is to move the logic from `ensureEntityType` into the pre/post hooks.

**Comment by jensschuppe** [comment]
Thanks, @colemanw!

Did you _not_ keep the `TODO`s for deleting entities of the to-be-deleted entity type and for deleting custom fields in custom groups extending the entity type on purpose? I can imagine other extensions wanting to subscribe via pre/post hooks for deleting all those. I'm not sure about the custom fields though - are they actively being deleted by the custom group deletion?

**Comment by colemanw** [comment]
@jensschuppe That's fine to add that back in. But I'm pretty sure that deleting a custom group via APIv4 will also delete custom fields and custom data within that group, so I don't think this extension needs to do anything further.

**Pull request merged**